libs/pilight/protocols/868/fs20.c

```diff
@@ -0,0 +1,288 @@
+
+/*static void createUnit(int unit) {
+	int binary[255];
+	int length = 0;
+	int i=0, x=0;
+
+	length = decToBinRev(unit, binary);
+	for(i=0;i<=length;i++) {
+		if(binary[i]==1) {
+			x=i*4;
+			createLow(x, x+3);
+		}
+	}
+}*/
+
+/*static void createId(int id) {
+	int binary[255];
+	int length = 0;
+	int i=0, x=0;
+
+	length = decToBinRev(id, binary);
+	for(i=0;i<=length;i++) {
+		if(binary[i]==1) {
+			x=i*4;
+			createLow(16+x, 16+x+3);
+		}
+	}
+}*/
+
+/*static void createState(int state) {
+	if(state == 0) {
+		createHigh(44,47);
+	}
+}*/
+
+/*static void createFooter(void) {
+	arctech_switch_old->raw[48]=(AVG_PULSE_LENGTH);
+	arctech_switch_old->raw[49]=(PULSE_DIV*AVG_PULSE_LENGTH);
+}*/
+
+static int createSync(void) {
+	createLow(0, 24);
+	createHigh(24, 26);
+	return 26;
+}
+
+static int createByte(int byte, int p) {
+	uint8_t mask = 0x80;
+	int par = 0;
+	
+	while(mask != 0) {
+		if((byte & mask) != 0) {
+			createHigh(p, p + 2);
+			par = !par;
+		}
+		else
+			createLow(p, p + 2);
+		
+		mask >>= 1;
+		p += 2;
+	}
+	
+	if (par != 0)
+		createHigh(p, p + 2);
+	else
+		createLow(p, p + 2);
+	p += 2;
+	
+	return p;
+}
+
+static int createHouseCode(int housecode, int p) {
+	p = createByte(housecode >> 8, p);
+	p = createByte(housecode & 0xFF, p);
+	return p;
+}
+
+static int createCode(struct JsonNode *code) {
+	int housecode = -1;
+	int address = -1;
+	int state = -1;
+	double itmp = -1;
+	int p = 0;
+
+	if(json_find_number(code, "housecode", &itmp) == 0)
+		housecode = (int)round(itmp);
+	if(json_find_number(code, "address", &itmp) == 0)
+		address = (int)round(itmp);
+	if(json_find_number(code, "off", &itmp) == 0)
+		state=0;
+	else if(json_find_number(code, "on", &itmp) == 0)
+		state=1;
+
+	if(housecode == -1 || address == -1 || state == -1) {
+		logprintf(LOG_ERR, "arctech_switch_old: insufficient number of arguments");
+		return EXIT_FAILURE;
+	} else if(housecode > 65353 || housecode < 0) {
+		logprintf(LOG_ERR, "fs20: invalid housecode range");
+		return EXIT_FAILURE;
+	} else if(address > 255 || address < 0) {
+		logprintf(LOG_ERR, "fs20: invalid address range");
+		return EXIT_FAILURE;
+	} else {
+		//createMessage(id, unit, state);
+		
+		p = createSync();
+		p = createHouseCode(housecode, p);
+		p = createByte(address, p);
+		//clearCode();
+		//createUnit(unit);
+		//createId(id);
+		//createState(state);
+		//createFooter();
+		fs20->rawlen = p;
+	}
+	return EXIT_SUCCESS;
+}
+
+static void printHelp(void) {
+	printf("\t -t --on\t\t\tsend an on signal\n");
+	printf("\t -f --off\t\t\tsend an off signal\n");
+	printf("\t -c --housecode=housecode\t\t\tcontrol a device with this housecode\n");
+	printf("\t -a --address=address\t\t\tcontrol a device with this address\n");
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void fs20Init(void) {
+	protocol_register(&fs20);
+	protocol_set_id(fs20, "fs20");
+	protocol_device_add(fs20, "fs20", "fs20");
+	fs20->devtype = SWITCH;
+	fs20->hwtype = RF868;
```

Comment:
Do you have a RF868 hardware module as well?

Reply:
yes